Add nearest-position capability to general field transfers

GiudGiud opened the pull request:
refs #23585 

this should address the Cardinal transfer needs
@aprilnovak

GiudGiud committed:
Add nearest positions parameter to general field transfers
and make older app-based system rely on positions as well
refs #23585

GiudGiud committed:
Use nearest positions for user object general field transfer
refs #23585

GiudGiud committed:
Use nearest positions for general field shape evaluation transfers
- regold because now we break when the source is not nearest to the same position as the target
- we used to just keep track of the distance no matter what, then compare (and valid values would win, even if further)
refs #23585

GiudGiud committed:
Rework NN to either use app-based or positions-based KDTrees
for finding the origin point
refs #23585

GiudGiud committed:
Add strong parameter inconsistency checking in relation to nearest-position algorithm
- try to guess when user wants to do extrapolation but forgot the bounding box
- refuse from_main_app with nearest_positions as it doesnt meach much for UOs
- refuse source_app_must_contain_point & nearest point as we dont know which app create the origin anymore (mixed app KDTrees for positions)
refs #23585

GiudGiud committed:
Add documentation for nearest position transfer
refs #23585

Fix documentation failures

GiudGiud committed:
Add test for nearest position transfer for UOs
refs #23585

GiudGiud committed:
Add test for siblings + nearest positions for UO to field transfer
refs #23585

GiudGiud committed:
Add test for NN + nearest positions
refs #23585

GiudGiud committed:
Add test for NN + siblings + nearest positions
refs #23585

GiudGiud committed:
Fix NN support
- use positions of apps to create trees with nodes at the right location
- exit from searching value if not closest to a postion
refs #23585

GiudGiud committed:
Make sure syntax page shows up well for Positions

GiudGiud committed:
Allow multiple coordinate systems (on per-block basis) in simulations with field transfers,
if coord_transform is off/skipped

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1517288/) on 26ba959 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24319/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1517326/) on 26ba959 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9587b2</th>
      <th colspan=4>#24319 26ba95</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.20%</td>
      <td>85.21%</td>
      <td>+0.01%</td>
      <td>97.76%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87713</td>
      <td>87784</td>
      <td>+71</td>
      <td>131</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15242</td>
      <td>15238</td>
      <td>-4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24319/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24319/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24319/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
-
-    if (is_nodal)
+    // Nest a loop on apps only for the Positions case. Regular case only looks at one app
+    unsigned int num_apps_to_consider = _nearest_positions_obj ? _from_problems.size() : 1;
```
```suggestion
    const unsigned int num_apps_to_consider = _nearest_positions_obj ? _from_problems.size() : 1;
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
-    if (is_nodal)
+    // Nest a loop on apps only for the Positions case. Regular case only looks at one app
+    unsigned int num_apps_to_consider = _nearest_positions_obj ? _from_problems.size() : 1;
+    for (unsigned int app_index = 0; app_index < num_apps_to_consider; app_index++)
```
```suggestion
    for (const auto app_index : make_range(num_apps_to_consider))

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
+    for (unsigned int app_index = 0; app_index < num_apps_to_consider; app_index++)
     {
-      for (const auto & node : from_mesh.getMesh().local_node_ptr_range())
+      unsigned int i_from;
+      if (!_nearest_positions_obj)
+        i_from = i_source;
+      else
+        i_from = app_index;
```
ternary and can add const

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
+                                   Moose::VarFieldType::VAR_FIELD_ANY);
+
+      System & from_sys = from_var.sys().system();
+      unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
+      // FEM type info
+      auto & fe_type = from_sys.variable_type(from_var.number());
+      bool is_nodal = fe_type.family == LAGRANGE;
```
```suggestion
      const unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
      // FEM type info
      const auto & fe_type = from_sys.variable_type(from_var.number());
      const bool is_nodal = fe_type.family == LAGRANGE;
```

lindsayad replied:
@roystgnr's favorite, `is_nodal` :smile:

GiudGiud replied:
cleaning this up in another PR

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
 
-        points[i_from].push_back(*node + _from_positions[i_from]);
-        auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
+          if (_nearest_positions_obj &&
+              !closestToPosition(i_source, *node + _from_positions[i_from]))
```
Pretty cool!

lindsayad replied:
From the conversation thread in the input file, which re-confused me, I'd like a comment here about what this is doing conceptually. Here is how I currently understand this:

Take two 1D apps from a MultiApp, which both span along the x-axis. Let's say app 1 is to the right of app 0. We draw a perpendicular cut line equidistant between the "positions" of the apps. What I understand this condition here as doing is disqualifying any nodes from app 0 that fall the right of the cut line, and any nodes from app 1 that fall to the left of the cut line. Is my interpretation correct?

GiudGiud replied:
No it should not be the positions of the app. It s the positions from the positions object. 

But you're right that this is used to disqualify nodes. 
The area closer to a given position than any other position in the object will be used to group (partition?) source points. Then when you look for a value for a target point, you find the position it is the closest to, and use those corresponding source points

I ll add comments in the code on top of doco

lindsayad replied:
> No it should not be the positions of the app. It s the positions from the positions object.

Yea I understand that the multiapp and the position are different objects and the distinction is important. But I guess that when I think of the positions system I think that there is an app in the multiapp object that is associated with a position  in the positions object. Is that fair?

lindsayad replied:
(yet here I am reviewing this ... 🤔 )

GiudGiud replied:
> Is that fair?

no it s not fair. That would have been nice.
Think of April's HTGR : one app doing heat transfer, 400k pebbles so 400k positions. 
And we need to be distinguishing these "nearest pebble" areas to make sure we dont send data from one pebble to another

GiudGiud replied:
however your thinking is correct for ShapeEvaluation and UO transfer. In those cases, the source of the data is the entire app, not a group of nodes. So the positions are essentially matched to one app.

But there's a need there too, for example for the 400k apps 400k pebble case. Positions let us use the nearest app to gather data from.

lindsayad replied:
Ok sorry I meant to indicate a one way relationship. When a multiapp is using the `positions_objects` parameter, then an app in the multiapp will be associated with a position from (one of) the positions object. Is that right?

> Think of April's HTGR : one app doing heat transfer, 400k pebbles so 400k positions.

I don't think I understand this yet. Ok you have one app modeling heat transfer physics for 400k pebbles. And so I can imagine a `Positions` in that input file for the 400k pebbles. But what data transfer is that `Positions` object controlling? Data exchange within that single input? Or with another multiapp?

lindsayad replied:
> Positions let us use the nearest app to gather data from.

Yea I get this. And that's very cool.

> a group of nodes

This is still what I need to understand. I should look at the tests you introduced in your last PR. I imagine that added a test in which a position is logically associated with a group of nodes (or elements) as opposed to an app in a multiapp

lindsayad replied:
so filling your kd trees you are looping over your `num_sources` in the outer loop and then the inner loop is over nodes so O(n_sources * n_nodes). What you'd ideally want is to loop over nodes and then do a KD tree search to figure out the closest position match as opposed to what seems to currently be a linear search over the positions. E.g. you want to do kd tree searches to fill your kd trees 😆 It seems like this kind of algorithm improvement would be really beneficial for when you go to something like 400k pebbles/400k positions

GiudGiud replied:
I can see KDTrees being created by Positions objects and used for getNearestPosition

however, at this stage, I think it s early for such an optimization.

Wrt to the ordering of the loop, if we do a loop on nodes, then we need to cache all the information on the entire node loop then create the KDTrees. It makes sense but there s a memory cost to that

lindsayad replied:
> then we need to cache all the information on the entire node loop then create the KDTrees.

I'm not sure I understand this, but yea I'm fine with not optimizing at this point

GiudGiud replied:
So you can’t form KD trees as you are looping on the nodes afaik. You have to take all the data relevant to each KDtree and use it all at once to create the tree. 

So if we loop on nodes rather than trees., we need to cache the datat for all the trees as we are looping on nodes

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
+            continue;
 
-        local_values[i_from].push_back((*from_sys.solution)(dof));
+          auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
```
```suggestion
          const auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
-        auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
-        points[i_from].push_back(elem->vertex_average() + _from_positions[i_from]);
-        local_values[i_from].push_back((*from_sys.solution)(dof));
+          auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
```
```suggestion
          const auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -157,9 +183,19 @@ MultiAppGeneralFieldNearestNodeTransfer::evaluateInterpValuesNearestNode(
 
-    // Loop on all source problems
-    for (MooseIndex(_from_problems.size()) i_from = 0; i_from < _from_problems.size(); ++i_from)
+    unsigned int num_sources;
+    if (!_nearest_positions_obj)
+      num_sources = _from_problems.size();
+    else
+      num_sources = _nearest_positions_obj->getPositions(/*initial=*/false).size();
```
ternary, add `const`

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -157,9 +183,19 @@ MultiAppGeneralFieldNearestNodeTransfer::evaluateInterpValuesNearestNode(
+      num_sources = _nearest_positions_obj->getPositions(/*initial=*/false).size();
+
+    // Loop on all sources
+    for (unsigned int i_from = 0; i_from < num_sources; ++i_from)
```
```suggestion
    for (const auto i_from : make_range(num_sources))
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:
```diff
@@ -193,19 +229,23 @@ MultiAppGeneralFieldNearestNodeTransfer::evaluateInterpValuesNearestNode(
       unsigned int num_equidistant_problems = 0;
 
-      for (MooseIndex(_from_problems.size()) i_from = 0; i_from < _from_problems.size(); ++i_from)
+      for (unsigned int i_from = 0; i_from < num_sources; ++i_from)
```
```suggestion
      for (const auto i_from : make_range(num_sources))
```

lindsayad commented on framework/include/transfers/MultiAppGeneralFieldTransfer.h:
```diff
@@ -152,6 +162,10 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
 
+  // Positions object to use to match target points and origin points as closest to the same
+  // Position
+  const Positions * _nearest_positions_obj;
```
```suggestion
  const Positions * const _nearest_positions_obj;
```

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -144,6 +156,29 @@ MultiAppGeneralFieldTransfer::MultiAppGeneralFieldTransfer(const InputParameters
+    _nearest_positions_obj = &_fe_problem.getPositionsObject("_created_for_" + name());
+  }
+
+  // Dont let users get wrecked by bounding boxes if it looks like they are trying to extrapolate
```
Love this comment :laughing:

lindsayad commented on framework/src/utils/MooseAppCoordTransform.C:
```diff
@@ -528,6 +528,10 @@ MultiAppCoordTransform::setDestinationCoordTransform(
 {
   _destination_app_transform = &destination_app_transform;
 
+  // Assume user knows what they are doing if they have multiple coordinate systems
```
```suggestion
  // Don't error check mismatching coordinate system types if we've been asked to skip coordinate collapsing since in that case the mismatch doesn't matter
```

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/main.i:
```diff
@@ -0,0 +1,107 @@
+  [out]
+    type = Exodus
+    overwrite = true
+    hide = 'to_sub to_sub_elem'
```
why hide these?

GiudGiud replied:
reduce file size! They are set by ICs, nothing interesting happens to them

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/main.i:
```diff
@@ -0,0 +1,107 @@
+    variable = from_main_elem
+  []
+
+  [from_sub]
+    type = MultiAppGeneralFieldNearestNodeTransfer
+    from_multi_app = sub
+    source_variable = to_main
+    variable = from_sub
+    use_nearest_position = input
+    bbox_factor = 100
+  []
```
So each node in the main application will only consider nodes from the multiapp whose position is closest to the main app node?

GiudGiud replied:
yes. Except it s not necessarily the app position that is used. It s whatever has been passed with use_nearest_pos

GiudGiud replied:
It s more like for Point A, let there be 2 nodes 1 and 2 and 2 positions X and Y
Point A is closest to node 1 and position Y
Node 1 is closest to position X
Node 2 is closest to position Y
We match point A with node 2 instead of 1, because being closest to the same position is more important

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/tests:
```diff
@@ -0,0 +1,36 @@
+      input = 'main.i'
+      cli_args = 'Transfers/from_sub/use_nearest_position=input Transfers/from_sub_elem/use_nearest_position=input'
+      exodiff = 'main_out.e main_out_sub0_out.e main_out_sub1_out.e'
+      detail = "for a 2D case with multiple subapps"
```
```suggestion
      detail = "for a 2D case with multiple subapps."
```

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/tests:
```diff
@@ -0,0 +1,36 @@
+  design = 'transfers/MultiAppGeneralFieldUserObjectTransfer.md'
+
+  [2d_overlay]
+    requirement = "The system shall be able to transfer spatial field using a nearest-node (source) "
```
```suggestion
    requirement = "The system shall be able to transfer a spatial field using a nearest-node (source) "
```

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/main_between_multiapp.i:
```diff
@@ -0,0 +1,97 @@
+
+[Transfers]
+  # Nodal to nodal variables
+  [app1_to_2_nodal_nodal]
+    type = MultiAppGeneralFieldNearestNodeTransfer
+    from_multi_app = ma1
+    to_multi_app = ma2
+    source_variable = sent_nodal
+    variable = received_nodal
+    use_nearest_position = input_app1
+    # slight inflation to avoid floating point issues on borders
+    bbox_factor = 1.000001
+    search_value_conflicts = true
+  []
```
And the same for the between mutliapp case? When looping over the "to" nodes, we will only consider a "from" node as a candidate if it is from the from-multiapp whose position is closest to the "to" node?

GiudGiud replied:
closest "to the same position" as the target node

lindsayad commented on test/tests/transfers/general_field/nearest_node/nearest_position/tests:
```diff
@@ -0,0 +1,36 @@
+      input = 'main.i'
+      cli_args = 'Transfers/from_sub/from_app_must_contain_point=true Transfers/from_sub_elem/from_app_must_contain_point=true'
+      expect_err = 'We do not support using both nearest positions matching and checking if target'
+      detail = 'both nearest positions algorithm and asking for origin points to be contained within an app domain are requested as this is not supported'
```
```suggestion
      detail = 'both nearest positions algorithm and asking for origin points to be contained within an app domain are requested as this is not supported.'
```

lindsayad commented on test/tests/transfers/general_field/shape_evaluation/regular/tests:
```diff
@@ -136,11 +140,15 @@
+      expect_err = "(Search for nearest position found several matches|multiple valid values from "
+                   "equidistant points were )"
       detail = "when using the closest app rather than the first valid value to chose a value"
+      method = !opt
```
why this restriction?

GiudGiud replied:
check on who's closest is only done in devel for nearest position.

I ve been thinking and I ll change that. NN is just too dangerous

GiudGiud replied:
has been changed

GiudGiud commented on framework/include/positions/Positions.h:
```diff
@@ -36,7 +36,7 @@ class Positions : public GeneralReporter
 
   /// Find the nearest Position for a given point
-  const Point & getNearestPosition(const Point & target) const;
+  const Point & getNearestPosition(const Point & target, bool initial = false) const;
```
oh this must have regressed from the previous PR you already said that

GiudGiud replied:
```suggestion
  const Point & getNearestPosition(const Point & target, bool initial) const;
```

GiudGiud committed:
Dont use time stepper to check if time integrator exists, refs #23653

GiudGiud committed:
Apply suggestions from code review
- more const
- dont use default argument for nearest position
- index-range for loops
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1524962/) on 79c3690 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24319/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24319/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c0775cd908a3b075515f73d6a18d7b31cb5170f0`

GiudGiud committed:
More const-ness
Dont say stuff is nodal by default

GiudGiud commented on framework/doc/content/source/transfers/MultiAppGeneralFieldTransfer.md:
```diff
@@ -85,6 +86,33 @@ Only the `MultiAppGeneralFieldTransfer` [!param](/Transfers/MultiAppGeneralField
+be used to match origin and target points. When specified with the
+[!param](/Transfers/MultiAppGeneralFieldNearestNodeTransfer/use_nearest_position) parameter, each target
+point of a transfer will only be matched with source entities that are closest to the same `Position` as the
+target point. The source entities are applications for:
```
better now?

lindsayad replied:
Yes excellent

GiudGiud committed:
Make check for equi-distant positions active in opt mode

GiudGiud committed:
Add more comments on nearest-position scheme

GiudGiud committed:
Fix conflict value detection for nearest position : we dont store the origin app id number

GiudGiud committed:
Add test for positions partitioning origin nodes from a single child app

GiudGiud committed:
Add test dedicated to finding indetermination in nearest position algorithm
with the search value conflict option

GiudGiud commented:
See new test
With NN we get the classic nearest node picture (first pic)
With nearest position, because of how the positions are moved, only sources from nodes that are closest to the same position as the target point are used (second pic)
![base_NN](https://github.com/idaholab/moose/assets/25037151/bbfead29-c863-4b5f-8239-56b33bd24c73)
![nearest_position](https://github.com/idaholab/moose/assets/25037151/16d5538f-4a0a-4363-afb9-3ac4d5c6de64)

added another commit to fix an issue with search value conflict + NN + positions

lindsayad commented:
I'm going to play around this afternoon and then we should be able to move ahead!

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -84,23 +84,19 @@ Positions::getNearestPosition(const Point & target, const bool initial) const
-#ifndef NDEBUG
-  // Check that no two positions are equidistant to the target
-  for (const auto i : index_range(_positions))
-    if (MooseUtils::absoluteFuzzyEqual((_positions[i] - target).norm_sq(), nearest_distance_sq) &&
-        (nearest_index != i))
+    // Check that no two positions are equidistant to the target
+    else if (MooseUtils::absoluteFuzzyEqual((_positions[i] - target).norm_sq(),
+                                            nearest_distance_sq))
       mooseWarning(
           "Search for nearest position found several matches: " + Moose::stringify(_positions[i]) +
               " and " + Moose::stringify(_positions[nearest_index]),
           " at a distance of " + std::to_string(std::sqrt(nearest_distance_sq)));
-#endif
```
Oh no never mind, https://github.com/idaholab/moose/pull/24196 is something else

GiudGiud commented:
Thanks for the reviews! 
@jortensi I was thinking of using this make sure data stays confined to each ABTR assembly
@aprilnovak @yukkiezy this should be able to differentiate between ducts and pins if you use a position for each pin center and duct wall center (so 6)
Please let me know if you hit any issue, user feedback is critical on these transfer things
